### PR TITLE
[java] Allow to customize target constructor args for EventFiringDecorator

### PR DESCRIPTION
**Pull request by mykola-mokhnach**
### Description

Potentially addresses https://github.com/appium/java-client/issues/1694#issuecomment-1281998055

I don't know how good the idea itself is, maybe it still needs some refactoring.
cc @valfirst 

### Motivation and Context
https://github.com/appium/java-client/issues/1694

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: java/src/org/openqa/selenium/support/decorators/WebDriverDecorator.java [hunk 1]
```diff
@@ -336,9 +346,22 @@ protected final <Z> Z createProxy(final Decorated<Z> decorated, Class<Z> clazz)
...
+    }
+
     try {
-      return proxy.newInstance();
-    } catch (ReflectiveOperationException e) {
+      Constructor<?> dstConstructor = proxy
+        .getConstructor(Arrays.stream(targetWebDriverConstructorArgs)
+          .map(Object::getClass)
+          .toArray(Class<?>[]::new)
+        );
```
Comment (valfirst) [inline java/src/org/openqa/selenium/support/decorators/WebDriverDecorator.java]: That might not work in 100% cases, for example: constructor arg type is `List`, but argument can be `ArrayList`. And for such cases fully automated search for applicable constructor is not a trivial task. So I would recommend to overload `WebDriverDecorator` constructor and add one more field:

```java
  private final Class<T> targetWebDriverClass;
  private final Class<?>[] targetWebDriverConstructorArgsTypes;
  private final Object[] targetWebDriverConstructorArgs;

  private Decorated<T> decorated;

...

  public WebDriverDecorator(Class<T> targetClass) {
    this(targetClass, new Object[0]);
  }

  public WebDriverDecorator(Class<T> targetClass, Object[] targetConstructorArgs) {
    this.targetWebDriverClass = targetClass;
    this.targetWebDriverConstructorArgs = targetConstructorArgs;
    this.targetWebDriverConstructorArgsTypes = Arrays.stream(targetWebDriverConstructorArgs)
          .map(Object::getClass)
          .toArray(Class<?>[]::new);
  }

  public WebDriverDecorator(Class<T> targetClass, Class<?>[] targetWebDriverConstructorArgsTypes, Object[] targetConstructorArgs) {
    this.targetWebDriverClass = targetClass;
    this.targetWebDriverConstructorArgsTypes = targetWebDriverConstructorArgsTypes;
    this.targetWebDriverConstructorArgs = targetConstructorArgs;
  ...
```

File: java/src/org/openqa/selenium/support/events/EventFiringDecorator.java [hunk 2]
```diff
@@ -162,6 +162,7 @@
   private final List<WebDriverListener> listeners;
 
   public EventFiringDecorator(WebDriverListener... listeners) {
+    super();
```
Comment (valfirst) [inline java/src/org/openqa/selenium/support/events/EventFiringDecorator.java]:
```suggestion
```
not needed, it's always invoked implicitly

File: java/src/org/openqa/selenium/support/decorators/WebDriverDecorator.java [hunk 3]
```diff
@@ -336,9 +356,22 @@ protected final <Z> Z createProxy(final Decorated<Z> decorated, Class<Z> clazz)
       .getLoaded()
       .asSubclass(clazz);
 
+    if (clazz.isInterface()
+        || this.targetWebDriverConstructorArgs.length == 0
+        || this.targetWebDriverConstructorArgTypes.length == 0
+        ) {
```
Comment (valfirst) [inline java/src/org/openqa/selenium/support/decorators/WebDriverDecorator.java]:
```suggestion
    if (clazz.isInterface()) {
```
the code below should work perfectly for constructors without args, so no need to add these checks

**Review by valfirst — COMMENTED** [review]
LGTM,
but also I believe there is some rule such bug fixes should be covered by unit tests.

**Comment by mykola-mokhnach** [comment]
> but also I believe there is some rule such bug fixes should be covered by unit tests.

I would be happy to do that if I knew how to run these tests locally. Is there some tutorial available on how to configure the IDE to be able to build/run the java selenium project and unit tests?

**Comment by pujagani** [comment]
@mykola-mokhnach Please have a look at https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md#step-5-test

**Comment by pujagani** [comment]
Thank you @mykola-mokhnach. The changes look good to me. Please add relevant tests and then I will merge the changes. Thank you!

**Review by valfirst — APPROVED** [review]

**Comment by mykola-mokhnach** [comment]
Unfortunately I have miserably failed to run neither the project nor tests locally. It looks like bazel is trying to download half of the internet and even after that it's not happy by something else. I was not able to also configure IntelliJ or VsCode with Bazel plugin installed to load the project and to build it properly, so I could have proper source autocomplete. Maybe I am still missing some docs...

Eventually, most of the code has been added in a completely "blind" manner with the hope CI could do its job.

**Comment by pujagani** [comment]
I appreciate your effort. Sorry to hear that setting it up was such a task. Usually, the bazel plugin makes it easy to import bazel project into Iintellij by importing the workspace file present in the folder https://github.com/SeleniumHQ/selenium. 

Thank you for adding the test! If there is an issue, I will take a look and help fix it.

**Comment by diemol** [comment]
I was sitting with @mykola-mokhnach a moment ago, and now he is able to run tests. Thanks for checking, @pujagani

**Comment by mykola-mokhnach** [comment]
I was able to resolve my issue with IDEA setup thanks to @diemol 
the trick was to use a separate bazel import flow rather than standard project import thing. Still not quite sure what was missing in VSCode, but I'm happy with what I have anyway. 

I was able to do more through local testing and it looks like several assumptions I did about ByteBuddy are wrong and the whole idea must be refined. I need to close the PR for further investigation.

**Comment by mykola-mokhnach** [comment]
Just to provide more context on this issue:

One of the main limitations with ByteBuddy is that it cannot patch an existing instance. It is necessary to create a proxy class first and then use its instance to invoke any driver calls. For example:

```

IOSDriver driver = new IOSDriver(caps);
...
WebDriverListener webDriverListener = new WebDriverListener();
IOSDriver proxy = (IOSDriver) new EventFiringDecorator(IOSDriver.class, webDriverListener).decorate(driver);
```

would create a `second` IOSDriver instance using the current implementation. 

This approach works for interfaces though, because by calling `decorate` we basically create a vanilla Object instance and then tell it to implement all methods (they all are public) of supported interfaces (e.g. have a dummy there). When any method of this proxy is being invoked we redirect the call to the original instance and trigger callbacks when necessary.

I need to verify if the above approach would be possible for classes as well

**Pull request closed without merging**